design/specs/navigation.md

```diff
@@ -0,0 +1,133 @@
+
+### Navigation via site header
+
+- The logo in the site header will lead visitors to the site's root, in this case the database homepage.
+- A breadcrumb-like navigation will be displayed at the top part of the site's header as the user navigates through the pages.
+- When users click on a breadcrumb element, a [page-selector](#page-selector) component is displayed with a list of schemas or entities based on the breadcrumb level.
+- Upon reaching the record level, the top navigation will show a 'Go to Record' action, which will open the [Record Selector](/engineering/specs/record-selector) component.
+- The current page will not be included in the breadcrumb path.
+- The current page label will be placed underneath the top header component in a sub-header component.
+- An icon will be used to identify each type of page (table, schema, record etc.)
+- When we eventually add support for several databases, we'll need CRUD UI for databases, and we may design some additional UI that stays within the header at that point.
+
+### Navigation via links within pages
+
+Throughout Mathesar, the following links exist within pages which allow for more targeted navigation to other pages.
+
+- **Database Page**
+  - Links to **Schema Pages** in a list of schemas.
+
+- **Schema Page**
+  - Links to **Table Pages** in a list of tables.
+  - Links to **Query Pages** in a list of queries.
+
+- **Table Page**
+  - Within the table data:
+    - Links to the **Record Page** for each row via the row's primary key cell value. _Note: Links within cell values will only be clickable when the cell is selected. This UX should be consistent with the behavior of the "URL" UI Type._
+    - Links to **Record Pages** for related records via the FK cell values
+  - Links to the **Table Page** for related tables via a section within the Table Inspector.
+
+- **Data Explorer**
+  - Features the same links as described by the "Within the table data" bullet point on Table Page above.
+
+- **Record Page**
+  - Links to other **Record Pages** via FK fields on directly on the record.
+  - Links to other **Record Pages** via tabular displays of _other records linking to this record_ via the same UI as the Table Page which has links in the PK cells.
+
+### Navigation via page url
+
+- **Home Page**:
+  - Route: `/`
+  - Note: the Home Page route responds with an HTTP 302 redirect to the Database Page for the database configured when installing Mathesar. When we eventually add support multiple databases, we may wish to give this design more thought.
+
+- **Database Page**:
+  - Route: `/{database_name}/`
+  - Favicon: Mathesar logo
+  - `<title>`: `{database_name} | Mathesar`
+
+- **Schema Page**:
+  - Route: `/{database_name}/{schema_id}`
+  - Favicon: Schema
+  - `<title>`: `{schema_name} | Mathesar`
+
+- **Table Page**:
+  - Route:  `/{database_name}/{schema_id}/{table_id}/`
+  - Favicon: Table
+  - `<title>`: `{table_name} | {schema_name} | Mathesar`
+
+- **Query Editor Page**:
```

Comment:
What is the query editor page?

Reply:
I wrote that text. Maybe this should be "Data Explorer" instead? My understanding is that we are currently implementing a page where the user edits the query definition (while also previewing the results of that query) -- and that in the "Query Page" feature, we plan to implement a page where users will view the results of the query (without the UI to edit the query definition). I think I've just been subconsciously categorizing all our pages as nouns and "Data Explorer" didn't quite fit into the pattern of "Database", "Schema", "Table", "Record" etc so I was thinking we'd have a "Query Editor Page" and a "Query Results Page" or something, just to keep the naming consistent. But I have no qualms about changing this to "Data Explorer" (and I'm happy to add a commit which makes this change).

Reply:
I would suggest three pages:

- **Query Page**: The is the "query results page" you're talking about. URL `/{database_name}/{schema_id}/query/{query_id}/`
- **Data Explorer**: This is where you start creating a new query. URL `/{database_name}/{schema_id}/data-explorer/`
- **Data Explorer with existing query**: This is where you open a query in the Data Explorer. URL `/{database_name}/{schema_id}/data-explorer/{query_id}/`

Reply:
After wrestling with this sub-thread a bit, I've realized that I _do_ have some qualms here, and that they're actually deeper than the URL choices.

But before I launch into my thinking, I'll begin by saying I don't feel very strongly about this and am fine with moving forward using your suggestion. These thoughts are out-of-scope with the Navigation specs anyway. But if my thinking resonates with you, then consider my proposal at the end of this message. If not, let's go with your proposal and I'll add a commit which modifies the spec to match it...

Fundamentally, I don't see the utility of using the terminology "query" _and_ the terminology "data explorer" to describe such similar aspects of the product. I understand that we use the data explorer to build queries. Are there _other_ things the data explorer will eventually build? If so, then distinct terms seems useful. If not, then I think the distinct terms will be confusing to users. These names have already been confusing _me_! I think users will have an easier time understanding Mathesar if we use _one_ of these terms consistently. While the term "data explorer" feels more inspiring and powerful, I prefer "query" because it feels simpler and more accurate.

The fact that I _"use the Data Explorer to create a Query"_ is not immediately evident from the terminology, whereas _"using the Query Editor to create a Query"_ is.

My inclination would be:

| Name | Url |
| -- | -- |
| New Query | /{database_name}/{schema_id}/query/ | 
| Query Editor | /{database_name}/{schema_id}/query/{query_id}/edit | 
| Query Viewer | /{database_name}/{schema_id}/query/{query_id}/ | 

With this terminology describing the features and functionality, we could retain the "explore" language in other contexts such as marketing copy like:

> Explore your data by creating custom queries and visualizations

Reply:
@seancolsen What you're proposing makes logic to me. When a query enters edit mode, users may be perplexed as to why the language changes to data explorer.
To maintain consistency with your example, we must also update the list with 'New Table,' 'New Record,' and 'New Schema.' Do you agree?

Reply:
> we must also update the list with 'New Table,' 'New Record,' and 'New Schema.' 

When you say "update the list", are you suggesting that we create additional pages for "New Table", "New Record", and "New Schema"? I'd be open to that, but I'm not sure it's necessary.

- When creating a schema, we only need a name. We acquire that via modal, which I thing provides for a more seamless workflow than a dedicated page would.
- We have two ways of creating a new table: (1) via import (which as @kgodey pointed out _does_ need an entry in the "Navigation via page url" section), or (2) via empty table (which I don't think warrants its own page).
- I think a dedicated page for "New Record" would be nice, but I'd rather not work on implementing it right now. I have some more ideas for this feature which I hope will aid with rapid data entry, but I think they're out of scope for this cycle, given that none of the storyboard screens seem to require a New Record page.

Reply:
I feel pretty strongly about continuing to use the term "Data Explorer", I think it's more friendly to new users and conveys the concept of what users are meant to do with the feature better. Also, users don't _have_ to save their output from the Data Explorer, they should be able to use it to explore without saving. So tying the name to only one potential output (a saved query) doesn't make sense.

If we rename anything (which I'm not proposing we do right now), I think we should get rid of the word "Query" because it's a technical term. We can't rename schemas or tables because they represent database objects, but queries are Mathesar-only objects and we can call them whatever we want.